### ISSUE TITLE
Search Head Clustering

### ISSUE THREAD

**Issue by cbowlesut** [jorritfolmer/puppet-splunk#3]
Very nice job with this module, I am currently using it to bootstrap our new Splunk environment.  Having some issues with our search head cluster, which is your example 7.  

I have sucessfully appled the SHC node configurations to a staging server and captured the puppet_* apps.
I have configured a deployer.
However, I am not clear on the best way to bootstrap new SHC nodes, since there is a chicken/egg problem with configuring them to join the cluster before pushing configurations.

Could you clarify the steps to take on the SHC nodes/slaves? 

Thanks!

**Comment by jorritfolmer** [comment]
Something like this should work:

1. stop Splunk on SHC nodes
1. manually put captured puppet_* apps in /opt/splunk/etc/apps on each SHC node
1. manually put captured puppet_* apps in /opt/splunk/etc/shcluster/apps on SH Deployer
1. make sure /opt/splunk/etc/system/local is empty on each SHC node
1. start SHC node
1. splunk init shcluster-config on each SHC node
1. splunk bootstrap shcluster-captain on one of the SHC nodes

The splunk init shcluster-config will create its configuration in /opt/splunk/etc/system/local individually for each node. So make sure to clean that between iterations of the steps above.

I've since discovered that it is possible to create 1 single configuration for SHC nodes that can copied to each SHC node and the SH Deployer. It involves pre-creating a `server_list = ` line with the GUID and fqdns of the SHC nodes. This belongs in the `[shcluster]` stanza of server.conf. It's basically the splunk init shcluster-config step in a more portable form. You can get the GUID from /opt/splunk/etc/instance.cfg on each SHC node.

This could be Puppetized, I think, however it will need custom facts, node dependencies and other Puppet wizardry.

**Comment by cbowlesut** [comment]
Awesome, thanks very much!

Yeah, I am trying to create a build process that will allow us to burn down and rebuild the entire infrastructure as needed:  for DR, testing, etc…
The search head cluster has definitely been the trickiest part.

We are using Ansible for our orchestration, so I think I will use your step-by-step using the full configuration string for “splunk init shcluster-config –option –option”.  For the time being,  I think it’s fine to have the SHC configs in /opt/splunk/etc/system/local on each server.

I am a little concerned about ongoing configuration management of the SHC nodes, since any pushed changes from the SHC Deployer will be placed into the default directory of the puppet_* apps (./puppet_*/default) and will place the current configs into the local directory.  This is designed to allow local modifications.  I might actually add an additional orchestration step to wipe out the local directories once the deployer is in control to avoid future conflicts.

Thanks again for the help!

~ Chris

**Comment by jorritfolmer** [comment]
Good point about the ongoing SHC node configuration from the SH Deployer! Currently puppet_*/local directories are created for the SHC stage node. This should indeed be changed to puppet_*/default. Thanks!

**Comment by jorritfolmer** [comment]
Ok I've pushed a new commit to address this. From my own testing it is less messy now. I've moved the mgmt_uri= to etc/system/local, so the puppet_* directories can remain stateless.
Also, on the search heads the configuration is only created if it doesn't exist. This makes the SH Deployer the leading entity again. Puppet is only used for initial config.
Only four steps need to taken, see README. Let me know what you think.
